crates/fluvio-spu/src/replication/leader/kv.rs

```diff
@@ -0,0 +1,367 @@
+    type Item = Result<Vec<u8>>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let next = match &mut self.inner_iter_fut {
+            InnerFutureState::Resolved(ref mut records) => records.pop(),
+
+            InnerFutureState::Pending(fut) => match ready!(fut.poll_unpin(cx)) {
+                Ok(records) if records.is_empty() => return Poll::Ready(None),
+                Ok(mut records) => {
+                    let next = records.pop();
+                    self.inner_iter_fut = InnerFutureState::Resolved(records);
+                    next
+                }
+                Err(e) => return Poll::Ready(Some(Err(e))),
+            },
+            InnerFutureState::None => {
+                let replica = self.replica.clone();
+                let last_read_offset = self.last_seen_offset;
+                let fut = Box::pin(Self::read_batch(replica, last_read_offset));
+                self.inner_iter_fut = InnerFutureState::Pending(fut);
+                cx.waker().wake_by_ref();
+                return Poll::Pending;
+            }
+        };
+        match next {
+            Some(record) => {
+                self.last_seen_offset = record.offset;
+                Poll::Ready(Some(Ok(record.record.into_value().into_vec())))
+            }
+            None => {
+                self.inner_iter_fut = InnerFutureState::None;
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+        }
+    }
+}
+
+enum InnerFutureState {
+    None,
+    Resolved(Vec<RecordItem>),
+    Pending(Pin<Box<dyn Future<Output = Result<Vec<RecordItem>>> + Send + 'static>>),
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{env::temp_dir, path::Path};
+
+    use fluvio_controlplane::replica::Replica;
+    use fluvio_kv_storage::KVStorage;
+    use fluvio_protocol::record::ReplicaKey;
+    use fluvio_storage::{config::ReplicaConfig, FileReplica};
+    use fluvio_types::PartitionId;
+    use flv_util::fixture::ensure_clean_dir;
+    use futures_util::StreamExt;
+
+    use crate::{
+        config::ReplicationConfig, control_plane::StatusMessageSink,
+        storage::SharableReplicaStorage,
+    };
+
+    use super::*;
+
+    #[fluvio_future::test]
+    async fn test_replica_kv() {
+        //given
+        let leader = create_replica("test_replica_kv").await;
+        let notifier = FollowerNotifier::shared();
+        let mut kv = LeaderKVStorage::<String, String, _>::new(LeaderReplicaLog::new(
+            leader.clone(),
+            notifier.clone(),
+        ));
+
+        //when
+        kv.put("key1", "value1").await.unwrap();
+        kv.put("key2", "value2").await.unwrap();
+        kv.put("key3", "value3").await.unwrap();
+
+        kv.delete(&"key2".to_string()).await.unwrap();
+
+        drop(kv);
+
+        let mut kv = LeaderKVStorage::<String, String, _>::new(LeaderReplicaLog::new(
```

Comment:
why is using LeaderKvStorage here.  That seems like integration testing.  Or least move to different folder.  Here, we just want to test `LeaderReplicaLog`

Reply:
We wrote `LeaderReplicaLog` for using it within KvStorage. So, this one test to check it works together.